src/System.Binary.Base64/System/Binary/Base64Decoder.cs

```diff
@@ -0,0 +1,266 @@
+        {
+            int i0 = srcBytes;
+            int i1 = Unsafe.Add(ref srcBytes, 1);
+            int i2 = Unsafe.Add(ref srcBytes, 2);
+
+            i0 = Unsafe.Add(ref decodingMap, i0);
+            i1 = Unsafe.Add(ref decodingMap, i1);
+            i2 = Unsafe.Add(ref decodingMap, i2);
+
+            i0 <<= 18;
+            i1 <<= 12;
+            i2 <<= 6;
+            
+            i1 |= i2;
+
+            i0 |= i1;
+            return i0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DecodePadByTwo(ref byte srcBytes, ref sbyte decodingMap)
+        {
+            int i0 = srcBytes;
+            int i1 = Unsafe.Add(ref srcBytes, 1);
+
+            i0 = Unsafe.Add(ref decodingMap, i0);
+            i1 = Unsafe.Add(ref decodingMap, i1);
+
+            i0 <<= 18;
+            i1 <<= 12;
+
+            i0 |= i1;
+            return i0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteThreeBytes(ref byte destBytes, int i0)
+        {
+            destBytes = (byte)(i0 >> 16);
+            Unsafe.Add(ref destBytes, 1) = (byte)(i0 >> 8);
+            Unsafe.Add(ref destBytes, 2) = (byte)i0;
+        }
+
+        public TransformationStatus Transform(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
+        {
+            ref byte srcBytes = ref source.DangerousGetPinnableReference();
+            ref byte destBytes = ref destination.DangerousGetPinnableReference();
+
+            int srcLength = source.Length / 4 * 4;  // only decode input up to the closest multiple of 4.
+            int destLength = destination.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+
+            if (source.Length == 0) goto DoneExit;
+
+            ref sbyte decodingMap = ref s_decodingMap[0];
+
+            while (sourceIndex < srcLength - 4) // TODO: you can make this more efficient by using unsafe pointer as the loop limit
+            {
+                int i0 = Decode(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
+                if (i0 < 0) goto InvalidExit;
+                if (destIndex > destLength - 3) goto DestinationSmallExit; // TODO: you can get rid off this line by including it in loop limit
+                WriteThreeBytes(ref Unsafe.Add(ref destBytes, destIndex), i0);
+                destIndex += 3;
+                sourceIndex += 4;
+            }
+
+            if (sourceIndex >= srcLength) goto NeedMoreExit;
+
+            int padding = 0;
+
+            if (Unsafe.Add(ref srcBytes, srcLength - 1) == s_encodingPad)
+            {
+                padding = (Unsafe.Add(ref srcBytes, srcLength - 2) == s_encodingPad) ? 2 : 1;
+            }
+
+            if (padding == 1)
+            {
+                int i0 = DecodePadByOne(ref Unsafe.Add(ref srcBytes, sourceIndex), ref decodingMap);
```

Comment:
Note that each of these if-else cases will end up having a lot of common code - each of them will fetch all 4 bytes (the last byte will be likely fetched twice - first for the comparison, and second time to do the actual work), map and shift the first two at least, and or them together. Lean and mean implementation would avoid this duplicated code.